bundle/artifacts/build.go

```diff
@@ -27,61 +29,146 @@ func (m *build) Name() string {
 func (m *build) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
 	var diags diag.Diagnostics
 
+	cacheDir, err := b.CacheDir(ctx)
+	if err != nil {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Warning,
+			Summary:  "Failed to set up cache directory",
+		})
+	}
+
 	for _, artifactName := range sortedKeys(b.Config.Artifacts) {
 		a := b.Config.Artifacts[artifactName]
 
-		if a.BuildCommand == "" {
-			continue
-		}
+		if a.BuildCommand != "" {
+			err := doBuild(ctx, artifactName, a)
+			if err != nil {
+				diags = diags.Extend(diag.FromErr(err))
+				break
+			}
 
-		cmdio.LogString(ctx, fmt.Sprintf("Building %s...", artifactName))
+			if a.Type == "whl" {
+				dir := a.Path
+				distPath := filepath.Join(a.Path, "dist")
+				wheels := python.FindFilesWithSuffixInPath(distPath, ".whl")
+				if len(wheels) == 0 {
+					diags = diags.Extend(diag.Errorf("cannot find built wheel in %s for package %s", dir, artifactName))
+					break
+				}
+				for _, wheel := range wheels {
+					a.Files = append(a.Files, config.ArtifactFile{
+						Source: wheel,
+					})
+				}
+			} else {
+				log.Warnf(ctx, "%s: Build succeeded", artifactName)
+			}
 
-		var executor *exec.Executor
-		var err error
-		if a.Executable != "" {
-			executor, err = exec.NewCommandExecutorWithExecutable(a.Path, a.Executable)
-		} else {
-			executor, err = exec.NewCommandExecutor(a.Path)
-			a.Executable = executor.ShellType()
-		}
+			// We need to expand glob reference after build mutator is applied because
+			// if we do it before, any files that are generated by build command will
+			// not be included into artifact.Files and thus will not be uploaded.
```

Comment:
They will be uploaded if they used in `libraries`section, right? The comment only matter when we have an artifact section but the artifact is not used later in the code?

Reply:
that's correct, as far as I can see (I did not write this comment, it already existed, but I agree with it).